중고물품 카테고리 등록 구현

junshock5 opened the pull request:
1. 카테고리 DB설계,  DTO, controller, service, Mapper 추가 
2. 카테고리명 등록시 ADMIN 계정일 경우에만 되게 변경
3. adminLoginCheck advice method 추가
4. account 테이블명 user테이블로 변경
5. User 로그인시 관리자 계정인지 확인되게 변경
6. ProductDTO에 categoryId, fileId 추가
7. 관리자 계정 세션명 LOGIN_ADMIN_ID으로 변경

junshock5 committed:
adminLoginCheck advice method 추가
카테고리 DTO, controller, service, Mapper 추가 및 등록 메서드 개발
account 테이블명 user테이블로 변경
User 로그인시 관리자 계정인지 확인되게 변경
ProductDTO에 categoryId, fileId 추가
관리자 계정 세션명 LOGIN_ADMIN_ID으로 변경

f-lab-dev commented on src/main/java/com/market/server/aop/LoginCheckAspect.java:
```diff
@@ -54,13 +54,37 @@ public Object loginCheck(ProceedingJoinPoint proceedingJoinPoint) throws Throwab
     }
+
+    @Around("@annotation(com.market.server.aop.LoginCheck) && @ annotation(loginCheck)")
+    public Object adminLoginCheck(ProceedingJoinPoint proceedingJoinPoint, LoginCheck loginCheck) throws Throwable {
```
굳이 메소드를 따로 만들어야할까요?
한 메소드에서 해결이 가능한데 그냥 복붙한 느낌입니다. 중복 코드가 있다면 메소드로 분리해주시고요~

junshock5 replied:
한 메소드에서 분기문으로 변경하였습니다.

f-lab-dev commented on src/main/java/com/market/server/controller/CategoryController.java:
```diff
@@ -0,0 +1,34 @@
+
+@Api(tags = {"4. categorys"})
+@RestController
+@RequestMapping("/categorys")
```
`categories` 오타같습니다~

junshock5 replied:
변경하겠습니다.

f-lab-dev commented on src/main/java/com/market/server/controller/CategoryController.java:
```diff
@@ -0,0 +1,34 @@
+    /**
+     * 중고물품 카테고리 등록 메서드.
+     */
+    @PostMapping("insertCategory")
```
계속 똑같은 리뷰가 반복되는데 `POST` 메소드에는 어떤 의미가 담겨있나요?

현재 이슈 관리 등 시간이 지나도 계속 비슷한 리뷰가 반복되고 있습니다. 어떤 것이든 도큐먼트를 꼼꼼하게 읽고 개념을 이해하지 않으면 멘토링의 효과가 적어집니다.

이 멘토링은 단순 자바하고 스프링을 사용하는 법만 공부하기만 하는 것이 아니라 어떻게 하면 전반적으로 프로젝트를 "잘" 할수 있을까가 목적입니다.

귀찮으시더라도 도큐먼트 읽기, 작명이나 컨벤션, 문서화 등 디테일에 신경써주세요~
단순히 코드량을 늘리는 것은 더 위로 올라가는데에 도움이 되지 않습니다. 그냥 기술 이것저것 얕게 사용해서 만든 프로젝트 10개보다 적은 기술이라도 깊게 공부하고 설계한 프로젝트 1개가 훨씬 퀄리티가 높다고 볼 수 있습니다

junshock5 replied:
도큐먼트를 읽고 api url 규칙에 따라 변경하겠습니다.

junshock5 replied:
참조: https://www.slideshare.net/Byungwook/rest-api-60505484?fbclid=IwAR04YHz7rZVVp70IZjrRK3YnYs3nfINK0gxD1pRwD_yqinAz2W0znN7hUVE

f-lab-dev commented on src/main/java/com/market/server/controller/UserController.java:
```diff
@@ -73,15 +73,16 @@ public void signUp(@RequestBody UserDTO userDTO) {
+        } else if (userInfo != null) {
             loginResponse = LoginResponse.success(userInfo);
-            SessionUtil.setLoginMemberId(session, id);
+            if (userInfo.getStatus().toString() == "ADMINE")
```
`ADMINE` 오타로 보입니다. 정상적으로 동작하는지 돌려보고 push해주세요~
PR에 올라오는 코드는 모두 정상적으로 동작하는 코드여야합니다.

그리고 이렇게 오타 문제가 생기지 않게 하려면 어떻게 해야할까요?

junshock5 replied:
ADMIN으로 변경하겠습니다.

junshock5 replied:
maven checkstyle 검사를 하면 될거같습니다.

f-lab-dev commented on src/main/java/com/market/server/dto/CategoryDTO.java:
```diff
@@ -0,0 +1,13 @@
+import lombok.ToString;
+
+@Getter
+@Setter
```
이것도 습관적으로 setter를 달고있는건 아닌지 계속 생각해주세요~

junshock5 replied:
category name 변경시에 사용할 예정입니다.

f-lab-dev commented on src/main/java/com/market/server/service/Impl/CategoryServiceImpl.java:
```diff
@@ -0,0 +1,31 @@
+    @Autowired
+    private CategoryMapper categoryMapper;
+
+    private static Logger logger = LogManager.getLogger(CategoryServiceImpl.class);
```
`@Log4j2` 어노테이션을 단 것과 이거의 차이는 무엇일까요?

junshock5 replied:
`Log4j2` 어노테이션은 Log4j2 라이브러리를 사용하는 것입니다.
LogManager를 사용하는것은 Log4j logging system을 사용하는것 입니다.

f-lab-dev replied:
1. 댓글을 다시면서 `log4j2`라는 유저를 태그하셨네요. 그런 일이 발생하지 않도록 어노테이션을 언급할때는 \`\`를 꼭 사용해주세요
2. 저는 두 말이 같은 말로 보이는데 혹시 코드상의 차이도 확인해보셨나요?
3. 추가적으로 `logger` 가 상수가 아니기 때문에 바뀔 위험이 있어보입니다~

junshock5 replied:
1. `` 사용하겠습니다.

2. 테스트 결과 
`@log4j2` 사용후 log.debug 와 logger=LogManager.getLogger(CategoryServiceImpl.class)
할당 후 logger.debug와 같이 동작 하였습니다. 
어노테이션 사용하여 가독성을 높이겠습니다.
혹시 내부적으로 다르게 동작하는게 있을까요??

3. final 키워드 추가하였습니다.

f-lab-dev commented on src/main/java/com/market/server/utils/SessionUtil.java:
```diff
@@ -42,8 +42,8 @@ public static void setLoginMemberId(HttpSession session, String id) {
-    public static String getLoginOwnerId(HttpSession session) {
-        return (String) session.getAttribute(LOGIN_OWNER_ID);
+    public static String getLoginAdminId(HttpSession session) {
+        return (String) session.getAttribute(LOGIN_ADMIN_ID);
```
`session.getAttribute(LOGIN_ADMIN_ID);`가 null을 리턴하면 어떻게 될까요?

junshock5 replied:
null 을 리턴하면 로그인 체크시 id를 가져오지 못합니다.

f-lab-dev replied:
예외는 발생하지 않나요?

junshock5 replied:
null을 리턴할 경우 
java.lang.NullPointerException 이 발생합니다.

f-lab-dev reviewed (commented):
이 브랜치에도 관리자 로그인 구현, 중고물품 카테고리 등록 여러 이슈가 함께 들어있습니다.
이슈는 최대한 잘게 분리해야합니다~

junshock5 committed:
1. LogManager.getLogger 로 할당 하지 않고 @Log4j2 사용하게 변경
2. API URL categorys -> categories 오타 수정
3. REST API rule 인 @Post 어노테이션에서 넣다는 뜻이 있음으로 insert url 제거 , @Patch update 제거
4. ADMINE -> ADMIN 오타 수정
5. UserDTO에 Status -> String type 수정
6. application.properties에 tomcat server.port 입력
7. adminLoginCheck 메서드를 삭제하고 loginCheck메서드에서 분기로 변경, proceedingJoinPoint.getArgs() for문 제거 후 인덱스로 set하게 변경

f-lab-dev commented on src/main/java/com/market/server/aop/LoginCheckAspect.java:
```diff
@@ -39,51 +38,30 @@
-        if (Id == null) {
-            logger.debug(proceedingJoinPoint.toString()+ "accountName :" + Id);
+        String id = null;
+        if (LoginCheck.UserType.ADMIN.equals(loginCheck.type()))
```
이럴땐 switch case문이 더 깔끔하지 않을까요?

f-lab-dev commented on src/main/java/com/market/server/aop/LoginCheckAspect.java:
```diff
@@ -39,51 +38,30 @@
-        }
-        return proceedingJoinPoint.proceed(modifiedArgs);
-    }
+    private static int idIndex = 0;
```
여기에도 thread safe하지 않은 변수로 추가하신 이유가 무엇일까요?

junshock5 replied:
0번째 파라미터인 accountID를 고정으로 사용하려고 넣었습니다.
생각해보니 static으로 쓸 이유는 없어 제거하겠습니다.

f-lab-dev commented on src/main/java/com/market/server/aop/LoginCheckAspect.java:
```diff
@@ -39,51 +38,30 @@
-            index++;
-        }
+        if(proceedingJoinPoint.getArgs()!=null)
+            modifiedArgs[idIndex] = id;
```
여기에는 제대로된 값이 들어갈 수 있을까요?

f-lab-dev commented on src/main/java/com/market/server/controller/CategoryController.java:
```diff
@@ -24,10 +24,10 @@ public CategoryController(CategoryServiceImpl categoryService) {
      * 중고물품 카테고리 등록 메서드.
      */
-    @PostMapping("insertCategory")
+    @PostMapping("categories")
```
URL 형식이 `/categories/categories` 이렇게 되는 것 같은데 어떻게 바꿔야할까요?

junshock5 replied:
/categories 입니다.

f-lab-dev commented on src/main/java/com/market/server/controller/UserController.java:
```diff
@@ -76,7 +76,7 @@ public void signUp(@RequestBody UserDTO userDTO) {
         } else if (userInfo != null) {
             loginResponse = LoginResponse.success(userInfo);
-            if (userInfo.getStatus().toString() == "ADMINE")
+            if (userInfo.getStatus().equals(UserDTO.Status.ADMIN.toString()))
```
enum은 이렇게 사용하는게 맞을까요?
`java enum best practice` 이런 식으로 검색해보시면 좋을 것 같네요~

junshock5 replied:
equals 메소드는 값 자체를 비교한다.
==는 주소값을 비교한다.
“==”연산자를 사용하면이 문제를 피할 수 있습니다.
참조: https://www.baeldung.com/a-guide-to-java-enums

junshock5 replied:
"=="연산자는 컴파일 타임 및 런타임 안전성을 제공합니다. 값이 null 인 경우 NullPointerException 이 발생하지 않습니다 .
반대로 equals 메소드가 사용되면 NullPointerException 이 발생합니다.

f-lab-dev commented on src/main/resources/application.properties:
```diff
@@ -11,4 +11,7 @@ spring.messages.encoding=UTF-8
+spring.redis.password=
+
+# tomcat
+server.port = 8888
```
여기에만 `=`에 띄어쓰기를 한 이유가 있으신가요?

junshock5 committed:
REST API URL 패턴을 restful하게 변경

junshock5 committed:
필요없는 전역변수 제거
가독성을 위해 if문을 case문으로 변경
잘못된 조건의 if문 변경
주소를 비교하는 "==" 비교문으로 변경
불필요한 띄어쓰기 제거

f-lab-dev commented on src/main/java/com/market/server/controller/ProductController.java:
```diff
@@ -46,7 +46,7 @@ public void registerProduct(String accountId, @RequestBody ProductDTO productDTO
      * 본인 중고물품 검색 메서드.
      */
-    @GetMapping("MyProducts")
+    @GetMapping("myProducts")
```
이럴땐 `my-products` 처럼 `-`를 활용하는게 좋습니다~

f-lab-dev commented on src/main/java/com/market/server/controller/UserController.java:
```diff
@@ -145,7 +145,7 @@ public void logout(HttpSession session) {
      * 회원 ID 삭제 메서드.
      */
-    @DeleteMapping("deleteID")
+    @DeleteMapping("")
```
`("")`는 생략이 가능합니다~

f-lab-dev commented on src/main/java/com/market/server/controller/UserProfileController.java:
```diff
@@ -24,20 +24,20 @@
-    @ApiOperation(value = "회원 단건 조회", notes = "userId로 회원을 조회한다")
-    @GetMapping(value = "/usermsg/{msrl}")
+    @ApiOperation(value = "회원 단건 언어별 조회", notes = "userId로 회원을 조회한다")
+    @GetMapping(value = "/users/{msrl}")
```
위의 `/users/{id}` 패턴과 겹치지 않을까요?

f-lab-dev commented on src/main/java/com/market/server/aop/LoginCheckAspect.java:
```diff
@@ -40,27 +40,34 @@
 @Log4j2
 public class LoginCheckAspect {
-    private static int idIndex = 0;
+    private int idIndex = 0;
```
이렇게 해도 결국 전역변수입니다. 여러 스레드가 이 값을 바꾸는데 각 스레드별로 온전한 결과를 만들어낼 수 있을까요?

junshock5 replied:
final 키워드로 `imutable`화 시켜야 할거같습니다.

f-lab-dev commented on src/main/java/com/market/server/controller/UserController.java:
```diff
@@ -76,7 +76,7 @@ public void signUp(@RequestBody UserDTO userDTO) {
         } else if (userInfo != null) {
             loginResponse = LoginResponse.success(userInfo);
-            if (userInfo.getStatus().equals(UserDTO.Status.ADMIN.toString()))
+            if (userInfo.getStatus() == (UserDTO.Status.ADMIN.toString()))
```
`userInfo.getStatus()`는 enum값인데 문자열과 `==` 비교가 가능한가요?

junshock5 replied:
equlas인 경우 값 비교이고 getStatus는 enum타입이므로 타입 안정성을위해 `==` 비교를 해야
런타임시까지 안전하게 npe가 발생하지 않습니다.

junshock5 committed:
Merge pull request #21 from f-lab-edu/feature/20

로그인 확인 코드 AOP 적용

junshock5 committed:
복붙으로 인한 메서드명 변경, restAPI url 변경

junshock5 committed:
LoginCheck시 사용되는 idIndex변수 지역변수로 변경
@PostMapping("") ->@PostMapping 변경
userDTO에 Status 타입 변경및 비교문 변경
userProfile controller에 중복되는 메서드 제거

f-lab-dev commented on src/main/resources/application.properties:
```diff
@@ -1,5 +1,5 @@
 # mysql
-spring.datasource.url=jdbc:mysql://localhost:3306/market?useUnicode=true&characterEncoding=utf8&serverTimezone=Asia/Seoul
+spring.datasource.url=jdbc:mysql://localhost:3307/market?useUnicode=true&characterEncoding=utf8&serverTimezone=Asia/Seoul
```
여기는 왜 포트를 다르게 주었나요? 3306이 일반적인 포트인데요

junshock5 replied:
회사에서 3306을 쓰고있어서
테스트 중에 들어간것 같습니다.
뺴고 커밋한다는게 3306으로 변경해놓겠습니다.

f-lab-dev commented on src/main/java/com/market/server/controller/UserProfileController.java:
```diff
@@ -23,19 +23,12 @@
-    @GetMapping("/users/{id}")
-    public SingleResult<UserDTO> getUserProfile(@ApiParam(value = "회원아이디", required = true) @PathVariable("id") String id) throws Exception {
+    @ApiOperation(value = "회원 한명 언어별 조회", notes = "userId로 회원을 조회한다")
+    @GetMapping(value = "/users/{id}")
```
이렇게하면 랜덤하게 userId를 넣어서 아무나 회원정보를 조회할 수 있지 않을까요?
권한관리가 필요해보입니다

junshock5 replied:
admin 계정 접근시 사용할수있게 변경하겠습니다.

f-lab-dev replied:
그러면 이 기능은 관리자를 위한 기능으로 보이는데요. 이 기능이 어떨 때 필요할까요? 관리자 기능은 관리자용 클래스에 모아두면 좋지 않을까요?

f-lab-dev replied:
이 기능은 꼭 필요한 기능인가요? 설명만 바뀌었네요

junshock5 committed:
userProfileController -> AdminUserController 변경 및 권한 설정, mysql port 3306 변경

junshock5 committed:
Update README.md

junshock5 committed:
Update README.md

junshock5 committed:
Update README.md

junshock5 committed:
Update README.md

f-lab-dev commented on src/main/java/com/market/server/controller/AdminUserController.java:
```diff
@@ -50,10 +56,17 @@
         return responseService.getSuccessResult();
     }
 
+    @DeleteMapping("SessionAll")
```
여기는 url이 대문자로 시작하는 이유가 무엇인가요?

f-lab-dev commented:
컨플릭트도 해결해주세요

junshock5 committed:
api url 소문자로 변경

junshock5 committed:
회원 조회 api 설명 변경

junshock5 committed:
Merge branch 'develop' into feature/18

f-lab-dev commented on src/main/java/com/market/server/controller/AdminUserController.java:
```diff
@@ -63,7 +63,7 @@ public CommonResult deleteUserProfile(String accountId, @ApiParam(value = "회
     }
 
-    @DeleteMapping("SessionAll")
+    @DeleteMapping("sessionAll")
```
보통 이럴때는 URL의 마지막에 행위를 명시합니다.
아래처럼 세션을 삭제한다는 의미로 표현하거나 마지막처럼 행위를 표시해주는 식으로도 할 수 있을 것 같네요

- `DELETE ~~/session`
- `DELETE ~~/session/clear`

f-lab-dev replied:
그리고 이 메소드는 단순히 로그아웃이네요. 로그아웃은 관리자용을 따로 만들 필요 없이 그냥 일반 사용자용 로그아웃을 쓰면 되지 않을까요?

junshock5 committed:
session 삭제 url 변경

junshock5 committed:
로그아웃 api UserController로 이동

f-lab-dev commented on src/main/java/com/market/server/controller/UserController.java:
```diff
@@ -161,6 +161,15 @@ public void logout(HttpSession session) {
+    /**
+     * 회원 로그아웃 메서드.
+     */
+    @DeleteMapping("session/clear")
```
사용자 입장에서는 `로그아웃`이라는 단어가 더 와닿을까요? 아니면 `세션 클리어`라는 단어가 더 와닿을까요?

f-lab-dev commented on src/main/java/com/market/server/controller/UserController.java:
```diff
@@ -161,6 +161,15 @@ public void logout(HttpSession session) {
+     * 회원 로그아웃 메서드.
+     */
+    @DeleteMapping("session/clear")
+    @LoginCheck(type = LoginCheck.UserType.USER)
```
이렇게 하면 관리자는 로그아웃을 할 수 없는걸까요?

f-lab-dev commented on src/main/java/com/market/server/controller/ProductController.java:
```diff
@@ -61,9 +61,24 @@ public ProductResponse myProductInfo(String accountId) {
     public void updateProducts(String accountId,
                                @PathVariable(name = "productId") int productId,
-                               @RequestBody ProductRequest productRequest) {
+                               @RequestBody ProductRequest PR) {
```
로컬변수인데 작명규칙은 지켜주세요

f-lab-dev commented on src/main/java/com/market/server/controller/ProductController.java:
```diff
@@ -61,9 +61,24 @@ public ProductResponse myProductInfo(String accountId) {
+                               @RequestBody ProductRequest PR) {
         UserDTO memberInfo = userService.getUserInfo(accountId);
 
+        ProductDTO productDTO = new ProductDTO(productId,
```
이렇게 길어질때는 빌더를 사용하면 어떤 곳에 어떤 값을 넣는건지 직관적으로 볼수 있습니다

junshock5 committed:
#18 productDTO 생성시 빌더 패턴 적용
관리자 로그아웃 추가
회원 로그아웃시 AOP적용
account -> user테이블 변경
product 필드 writerId -> accountID 변경

f-lab-dev commented on src/main/java/com/market/server/controller/AdminUserController.java:
```diff
@@ -62,4 +62,14 @@ public CommonResult deleteUserProfile(String accountId, @ApiParam(value = "회
+     * 관리자 로그아웃 메서드.
+     */
+    @DeleteMapping("")
+    @LoginCheck(type = LoginCheck.UserType.ADMIN)
```
이런거는 중복된 메소드를 하나 더 추가하는것보다는 `@LoginCheck`라는 어노테이션에 여러 권한을 넣을 수 있도록 하면 어떨까요?

junshock5 replied:
중복메서드를 삭제하고
로그아웃시 전체 세션정보를 삭제하게 변경하였습니다.

f-lab-dev commented on src/main/java/com/market/server/controller/AdminUserController.java:
```diff
@@ -62,4 +62,14 @@ public CommonResult deleteUserProfile(String accountId, @ApiParam(value = "회
+    /**
+     * 관리자 로그아웃 메서드.
+     */
+    @DeleteMapping("")
```
마찬가지로 이거는 세션을 delete한다는 관점으로 보면 안됩니다. 로그아웃이라는 사용자 관점의 단어를 명확하게 사용해주세요~

junshock5 replied:
로그아웃은 데이터를 delete하는게 아니고 상태를 수정하는거로 이해했습니다.
응용 프로그램 자체에 아무런 영향을주지 않기 때문에 delete는 어울리지 않고
http method중 PUT: 요청된 자원을 수정(UPDATE) 하는 것.
결론적으로 PUT이 적합한것 같습니다.

junshock5 committed:
#18 중고물품 카테고리 등록 구현
- 중복된 관리자 로그아웃 메서드 삭제
- 로그아웃시 전체 세션 삭제되게 변경
- 나의 정보 확인시 관리자 계정도 확인되게 변경
